x-pack/plugins/notifications/server/plugin.ts

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type {
+  PluginInitializerContext,
+  CoreSetup,
+  CoreStart,
+  Plugin,
+  Logger,
+} from '@kbn/core/server';
+import type {
+  NotificationsPluginSetupDeps,
+  NotificationsPluginStartDeps,
+  NotificationsPluginSetup,
+  NotificationsPluginStart,
+} from './types';
+import { EmailService } from './services';
+import { NotificationsConfigType } from './config';
+import { getResolvablePromise, type ResolvablePromise } from '../common/lib';
+import { PLUGIN_ID } from '../common';
+
+export class NotificationsPlugin
+  implements Plugin<NotificationsPluginSetup, NotificationsPluginStart>
+{
+  private readonly logger: Logger;
+  private readonly initialConfig: NotificationsConfigType;
+  private email: ResolvablePromise<EmailService>;
+  private emailConnector: string;
+
+  constructor(initializerContext: PluginInitializerContext<NotificationsConfigType>) {
+    this.logger = initializerContext.logger.get();
+    this.email = getResolvablePromise();
+    this.initialConfig = initializerContext.config.get();
+  }
+
+  public setup(core: CoreSetup, plugins: NotificationsPluginSetupDeps) {
+    // TODO this must be defaulted to 'elastic-cloud-email' for cloud
+    if (!plugins.actions) {
+      this.email.doReject(`Error starting notification services: 'actions' plugin not available.`);
+      return { email: this.email };
+    }
+
+    const emailConnector = this.initialConfig.connectors?.default?.email;
+    if (!emailConnector) {
+      this.email.doReject('Error starting notification services: Email connector not specified');
+      return { email: this.email };
+    }
+
+    if (!plugins.actions.isPreconfiguredConnector(emailConnector)) {
+      this.email.doReject(
+        `Error starting notification services: Unexisting email connector '${emailConnector}' specified`
+      );
+      return { email: this.email };
+    }
+
+    plugins.actions.registerUnsecuredActionsClientAccess(PLUGIN_ID);
+
+    this.emailConnector = emailConnector;
+
+    return {
+      email: this.email,
+    };
+  }
+
+  public start(core: CoreStart, plugins: NotificationsPluginStartDeps) {
+    if (this.emailConnector) {
+      plugins.actions.getUnsecuredActionsClient().then(
+        (actionsClient) => {
+          const email = new EmailService(PLUGIN_ID, this.emailConnector, actionsClient);
+          this.email.doResolve(email);
+        },
+        (error) => {
+          this.logger.warn(`Error starting notification services: ${error}`);
+          this.email.doReject(error);
+        }
+      );
+    }
+
+    return {
+      email: this.email,
+    };
+  }
```

Comment:
Two things I would like to suggest here:

1. From what I understand the email service is *async* only to handle the case of connectors not being available and to be able to store a specific error message in that promise. I'm not sure that is a good reason to make the service *async*, would be still nice to keep it *sync*.
2. All the business logic here (about 50 lines of code) is specific to the `EmailService`, there will be other services here in the future, to make it self contained and future proof, would be nice to remove this `EmailService` construction code out of the `plugin.ts` file.


Something like:

```ts
  private email?: EmailService;

  public setup(core: CoreSetup, plugins: NotificationsPluginSetupDeps) {
      this.email = new EmailService(core, plugins.actions);
      return { email: this.email }
  }

  public start() {
    this.email._boot();
    return { email: this.email! };
  }
```

where:

```ts
class EmailService {
  public error?: Error;

  constructor (protected readonly core, protected readonly actions) {
    this.boot();  
  }

  public _boot(): void {
    // ...
  }

  private assertIsReady() {
    // ...
  }

  public async sendPlainTextEmail(options) {
    await this.assertIsReady();
    // send email ...
  }
}
```

---

Another option could be to create a "provider" for the email service, where in the future we could use it to setup a non-connector based email sender:

```ts
  private emailProvider?: EmailServiceProvider;

  public setup(core: CoreSetup, plugins: NotificationsPluginSetupDeps) {
      this.emailProvider = new EmailServiceProvider();
      if (core.email) {
        this.emailProvider.setImplementation(new CoreEmailService(core));
      } else if (plugins.cloud) {
        this.emailProvider.setImplementation(new CloudEmailService(core, plugins.cloud));
      } else if (plugins.actions) {
        this.emailProvider.setImplementation(new AlertingEmailService(core, plugins.actions));
      }
      return { email: () => this.emailProvider.create() }
  }

  public start() {
      this.emailProvider._boot();
      return { email: () => this.emailProvider.create() }
  }
```

Reply:
Thanks for the suggestions!

(1) The email service is actually `async` because the _Actions_ plugin's `getUnsecuredActionsClient()` method is also async, so I cannot instantiate the `EmailService` until I have an instance of the `ActionsClient`. That being said, looking at the _Actions_ plugin code, I don't see a strong reason why this method is async, so perhaps we can simply make it synchronous. @ymao1 WDYT?
Another reason is, if we want to expose it on the `setup()` contract, we have to await for _Actions_ plugin `start()` contract in order to get the `ActionsClient`, but again, I don't know if there's a good reason not to expose the `getUnsecuredActionsClient()` in the `setup()`...

(2) Good point, I'll create a _factory_ and extract this logic there.

Reply:
@gsoldevila I can make `getUnsecuredActionsClient()` synchronous, you are correct that there doesn't seem to be a reason for it to be asynchronous, however, it uses `core.savedObjects.createInternalRepository` which is only available on `SavedObjectsServiceStart`, not on `SavedObjectsServiceSetup`, so I need to keep that function inside the `start`

Reply:
@gsoldevila It works also for us. The cases client, which is used in our routes, is initialized on `start`.

Reply:
@gsoldevila nice, LGTM, thx